katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/AbstractJpaQueryImpl.java

```diff
@@ -0,0 +1,192 @@
+		this.clazz = (Class<T>) meta.getImplementationClass();
+
+		this.parentEntityClass = entityClass;
+		this.parentAttr = attrMeta;
+		this.parentIds = entityIds;
+	}
+
+	@Override
+	public JpaQuery<T> setEnsureTotalOrder(boolean ensureTotalOrder) {
+		this.ensureTotalOrder = ensureTotalOrder;
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> addFilter(FilterSpec filters) {
+		this.filterSpecs.add(filters);
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> addSortBy(Direction dir, String... path) {
+		this.sortSpecs.add(new SortSpec(Arrays.asList(path), dir));
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> addSortBy(List<String> attributePath, Direction dir) {
+		this.sortSpecs.add(new SortSpec(attributePath, dir));
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> addSortBy(SortSpec order) {
+		this.sortSpecs.add(order);
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> setDefaultJoinType(JoinType joinType) {
+		this.defaultJoinType = joinType;
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> setJoinType(JoinType joinType, String... path) {
+		joinTypes.put(meta.resolvePath(Arrays.asList(path)), joinType);
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> setAutoGroupBy(boolean autoGroupBy) {
+		this.autoGroupBy = autoGroupBy;
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> setDistinct(boolean distinct) {
+		this.autoDistinct = false;
+		this.distinct = distinct;
+		return this;
+	}
+
+	@Override
+	public JpaQuery<T> addFilter(String attrPath, FilterOperator filterOperator, Object value) {
+		return addFilter(Arrays.asList(attrPath.split("\\.")), filterOperator, value);
+	}
+
+	@Override
+	public JpaQuery<T> addFilter(List<String> attrPath, FilterOperator filterOperator, Object value) {
+		addFilter(new FilterSpec(attrPath, filterOperator, value));
+		return this;
+	}
+
+	public List<SortSpec> getSortSpecs() {
+		return sortSpecs;
+	}
+
+	public boolean getEnsureTotalOrder() {
+		return ensureTotalOrder;
+	}
+
+	public JoinType getJoinType(MetaAttributePath path) {
+		JoinType joinType = joinTypes.get(path);
+		if (joinType == null)
+			joinType = defaultJoinType;
+		return joinType;
+	}
+
+	public VirtualAttributeRegistry getVirtualAttrs() {
+		return virtualAttrs;
+	}
+
+	public MetaDataObject getMeta() {
+		return meta;
+	}
+
+	@Override
+	public Class<T> getEntityClass() {
+		return clazz;
+	}
+
+	@Override
+	public AbstractQueryExecutorImpl<T> buildExecutor() {
+		B backend = newBackend();
+
+		@SuppressWarnings({ "rawtypes", "unchecked" })
+		QueryBuilder executorFactory = new QueryBuilder(this, backend);
+		executorFactory.applyFilterSpec();
+		executorFactory.applySortSpec();
+		int numAutoSelections = executorFactory.applyDistinct();
+
+		return newExecutor(backend, numAutoSelections);
+	}
+
+	protected abstract AbstractQueryExecutorImpl<T> newExecutor(B ctx, int numAutoSelections);
+
+	protected abstract B newBackend();
+
+	public List<?> getParentIds() {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove usage of generic wildcard type. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1452)